tools/xdp/s2n-quic-xdp/src/task/completion_to_tx/assign.rs

```diff
@@ -0,0 +1,240 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::if_xdp::UmemDescriptor;
+
+/// Trait to define how descriptors are assigned to TX workers
+///
+/// As the Completion ring is global for the entire socket, it is up to the application to decide
+/// which TX queues get which descriptors. This trait takes in a descriptor and decides if it
+/// pertains to a worker index or not.
+pub trait Assign: Unpin {
+    fn assign(&self, desc: UmemDescriptor, idx: u64) -> bool;
+}
+
+impl Assign for () {
+    #[inline]
+    fn assign(&self, _desc: UmemDescriptor, idx: u64) -> bool {
+        debug_assert_eq!(
+            idx, 0,
+            "assignment mode should only be used for single queue workflows"
+        );
+
+        // only assign descriptors to the first worker
+        idx == 0
+    }
+}
+
+/// Assignment strategy that is generic over framing and index alignment
+pub struct AssignGeneric<F: FrameToIndex, I: IndexToQueue> {
+    pub frame: F,
+    pub index: I,
+}
+
+impl<F: FrameToIndex, I: IndexToQueue> Assign for AssignGeneric<F, I> {
+    #[inline]
+    fn assign(&self, desc: UmemDescriptor, idx: u64) -> bool {
+        let v = self.frame.frame_to_index(desc);
+        let v = self.index.index_to_queue(v);
+        v == idx
+    }
+}
+
+/// Converts a frame address into a frame index
+pub trait FrameToIndex: Unpin {
+    fn frame_to_index(&self, desc: UmemDescriptor) -> u64;
+}
+
+pub struct AlignedFrame {
+    shift: u32,
+}
+
+impl AlignedFrame {
+    pub fn new(frame_size: u32) -> Self {
+        debug_assert!(frame_size.is_power_of_two());
+
+        let shift = frame_size.trailing_zeros();
+
+        debug_assert_eq!(
+            frame_size,
+            2u32.pow(shift),
+            "computing the square root of a power of two is counting the trailing zeros"
+        );
+
+        Self { shift }
```

Comment:
The idea here is if we know that the frame_size is a power of two, it's much cheaper to shift bits than perform regular integer division.